velox/docs/develop/operators.rst

```diff
@@ -291,6 +292,125 @@ Note: Presto allows multiple global grouping sets in a single SQL query.
 
 Hence, globalGroupingSets is a vector of groupIds.
 
+.. _ExpandNode:
+
+ExpandNode
+~~~~~~~~~~~
+
+For each input row, generates N rows with M columns according to specified 'projections'.
+'projections' is an N x M matrix of expressions: a vector of N rows each having M columns.
+Each expression is either a column reference or a constant. Both null and non-null constants are allowed.
+'names' is a list of M new column names. The semantic of this operator matches Spark. Using project and unnest can be
+employed to implement the expand functionality. However, the performance is suboptimal when creating an array
+constructor within the Project operation.
+
+.. list-table::
+   :widths: 10 30
+   :align: left
+   :header-rows: 1
+
+   * - Property
+     - Description
+   * - projections
+     - A vector of N rows each having M columns. Each expression is either a column reference or a constant.
+   * - names
+     - A list of new column names.
+
+ExpandNode is typically used to compute GROUPING SETS, CUBE, ROLLUP and COUNT DISTINCT.   
+
+To illustrate how ExpandNode works lets examine the following SQL query:
+
+.. code-block:: sql
+
+  SELECT l_orderkey, l_partkey, sum(l_suppkey) FROM lineitem GROUP BY ROLLUP(l_orderkey, l_partkey);
+
+In the planning phase, Spark generates an Expand operator with the following projection list:
+
+.. code-block::
+
+  [l_suppkey, l_orderkey, l_partkey, 0],
+  [l_suppkey, l_orderkey, null,      1],
+  [l_suppkey, null,       null,      3]
+
+Note: The grouping_id_0 in Spark is calculated based on a bitmask. If a certain column is selected, the bit value is assigned as 0; otherwise, it is assigned as 1. Therefore, the binary representation of the first row is (000), resulting in 0. The binary representation of the second row is (001), resulting in 1. The binary representation of the third row is (011), resulting in 3.
+
+For example, if the input rows are:
+
+.. code-block::
+
+  l_suppkey l_orderkey l_partkey
+  93        1          673
+  75        2          674
+  38        3          22
+
+After the computation by the ExpandNode, each row will generate 3 rows of data. So there will be a total of 9 rows:
+
+.. code-block::
+
+  l_suppkey l_orderkey l_partkey grouping_id_0 
+  93        1          673       0
+  93        1          null      1
+  93        null       null      3
+  75        2          674       0
+  75        2          null      1
+  75        null       null      3
+  38        3          22        0
+  38        3          null      1
+  38        null       null      3
+
+Then, these 9 rows of data are grouped by (l_orderkey, l_partkey, grouping_id_0), and the result of calculating l_orderkey, l_partkey, count(l_suppkey) is:
+
+.. code-block::
+
+  l_orderkey l_partkey count(l_suppkey)
```

Comment:
add group ID column

Reply:
@mbasmanova The group id column will be deleted after aggregation.

Reply:
That's right, but it needs to be clarified. It might be helpful to show the full query plan and explain that ExpandNode is followed by AggregationNode and by ProjectNode, then describe what each node does.